service-proxy.js

```diff
@@ -950,7 +949,44 @@ function reapSinglePeer(hostPort) {
         self.deletePeerIndex(serviceName, hostPort);
     }
 
-    self.channel.peers.delete(hostPort);
+    // TODO: info log/stat
+
+    peer.drain({
+        reason: 'reaped for expired advertisement',
+        direction: 'both',
+        timeout: self.drainTimeout
+    }, disconnectDrainDone);
+
+    function disconnectDrainDone(err) {
+        if (err &&
+            err.type === 'tchannel.drain.peer.timed-out') {
+            // TODO: stat?
+            self.logger.warn('forcibly closing reaped peer', self.extendLogInfo({
+                timeout: err.timeout,
+                elapsed: err.elapsed
+            }));
+            err = null;
+        }
+        thenCloseIt(err);
+    }
+
+    function thenCloseIt(err) {
+        if (err) {
+            self.logger.warn('error draining reaped peer, force closing', self.extendLogInfo({
+                error: err
+            }));
+        }
+        peer.close(thenDeleteIt);
+    }
+
+    function thenDeleteIt(err) {
+        if (err) {
+            self.logger.warn('error closing reaped peer, deleting it anyhow', self.extendLogInfo({
+                error: err
+            }));
+        }
+        self.channel.peers.delete(hostPort);
```

Comment:
There are kinds of annoying races that can happen.

We need to fix:

 - https://github.com/uber/tchannel-node/blob/master/connection.js#L477
 - https://github.com/uber/tchannel-node/blob/master/peer.js#L493

It's probably fine but could be "nicer" tm.